models/itsi_obj.go

```diff
@@ -0,0 +1,730 @@
+package models
+
+// Models and support for interacting (generically) with Splunk ITSI's
+// ITOA object-related REST APIs.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/tivo/terraform-provider-splunk-itsi/provider/util"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	resourceHashField      = "_tf_hash"
+	asyncUpdateCheckPeriod = 15 * time.Second
+	updateSuccessTimeout   = 100 * time.Second //If the update success cannot be confirmed within this period, the update will be considered as failed and retried.
+)
+
+var RestConfigs map[string]restConfig
+
+type RawJson json.RawMessage
+
+func (rj RawJson) ToInterfaceMap() (m map[string]interface{}, err error) {
+	var by []byte
+	if by, err = json.RawMessage(rj).MarshalJSON(); err != nil {
+		return
+	}
+	err = json.Unmarshal(by, &m)
+	return
+}
+
+func (rj RawJson) MarshalJSON() ([]byte, error) {
+	return json.RawMessage(rj).MarshalJSON()
+}
+
+func (rj *RawJson) UnmarshalJSON(data []byte) error {
+	*rj = RawJson(append(make([]byte, 0, len(data)), data...))
+	return nil
+}
+
+var GenerateResourceKey = func() (string, error) {
+	return uuid.GenerateUUID()
+}
+
+func init() {
+	err := yaml.Unmarshal([]byte(metadataConfig), &RestConfigs)
+	if err != nil {
+		log.Fatalf("Unmarshal: %v", err)
+	}
+}
+
+type restConfig struct {
+	RestInterface          string `yaml:"rest_interface"`
+	ObjectType             string `yaml:"object_type"`
+	RestKeyField           string `yaml:"rest_key_field"`
+	TFIDField              string `yaml:"tfid_field"`
+	MaxPageSize            int    `yaml:"max_page_size"`
+	GenerateKey            bool   `yaml:"generate_key"`
+	UnimplementedFiltering bool   `yaml:"unimplemented_filtering"`
+}
+
+type ItsiObj struct {
+	base Base
```

Comment:
Embed this please, so that we can easier access its exported fields/functions and don't have to do `.base.` all the time 🙏 Same in the Collection struct.
```suggestion
	Base
```